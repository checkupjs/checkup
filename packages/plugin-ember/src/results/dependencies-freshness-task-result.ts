import { BaseTaskResult, TaskResult, ui } from '@checkup/core';
import { DepFreshnessInfo } from '../types';

export default class DependenciesFreshnessTaskResult extends BaseTaskResult implements TaskResult {
  depFreshnessInfo!: DepFreshnessInfo;

  stdout() {
    ui.styledHeader(this.meta.taskName);
    ui.blankLine();

    this._writeFreshnessTable(this.depFreshnessInfo);
    ui.blankLine();
    // Todo: should look into color schemes
  }

  json() {
    return {
      meta: this.meta,
      result: {
        depFreshnessInfo: this.depFreshnessInfo,
      },
    };
  }

  pdf() {
    return undefined;
  }

  _writeFreshnessTable(dependencies: DepFreshnessInfo) {
    if (dependencies.tableBody && dependencies.tableBody.length === 0) {
      return;
    }

    ui.table(this._transformToTableData(dependencies.tableBody), {
      package: {},
      current: {},
      wanted: {},
      latest: {},
      packageType: {},
      url: {},
    });
  }

  _transformToTableData(data: Array<String[]>) {
    let result: {
      package: String;
      current: String;
      wanted: String;
      latest: String;
      packageType: String;
      url: String;
    }[] = [];
    // Todo: should look into making it cleaner here
    data.forEach(dependency => {
      const row = {
        package: dependency[0],
        current: dependency[1],
        wanted: dependency[2],
        latest: dependency[3],
        packageType: dependency[4],
        url: dependency[5],
      };
      result.push(row);
    });
    return result;
  }
}
