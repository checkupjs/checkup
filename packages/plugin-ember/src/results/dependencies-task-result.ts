import { BaseTaskResult, Task, TaskResult, toPairs, ui } from '@checkup/core';

import { PackageJson } from 'type-fest';

export default class DependenciesTaskResult extends BaseTaskResult implements TaskResult {
  emberLibraries!: PackageJson.Dependency;
  emberAddons!: Record<string, PackageJson.Dependency>;
  emberCliAddons!: Record<string, PackageJson.Dependency>;

  constructor(task: Task) {
    super(task);
    this.emberLibraries = {};
  }

  get hasDependencies() {
    return (
      Object.keys(this.emberLibraries).length ||
      [this.emberAddons, this.emberCliAddons].every(addons => {
        return Object.keys(addons.dependencies).length && Object.keys(addons.devDependencies);
      })
    );
  }

  stdout() {
    if (!this.hasDependencies) {
      return;
    }
    ui.styledHeader(this.meta.friendlyTaskName);
    this._writeDependencySection('Dependencies - Core Libraries', this.emberLibraries);
    this._writeDependencySection('Dependencies - Ember Addons', this.emberAddons.dependencies);
    this._writeDependencySection(
      'Dev Dependencies - Ember Addons',
      this.emberAddons.devDependencies
    );
    this._writeDependencySection(
      'Dependencies - Ember CLI Addons',
      this.emberCliAddons.dependencies
    );
    this._writeDependencySection(
      'Dev Dependencies - Ember CLI Addons',
      this.emberCliAddons.devDependencies
    );
  }

  json() {
    return {
      meta: this.meta,
      result: {
        emberLibraries: this.emberLibraries,
        emberAddons: this.emberAddons,
        emberCliAddons: this.emberCliAddons,
      },
    };
  }

  pdf() {}

  _writeDependencySection(header: string, dependencies: PackageJson.Dependency) {
    if (Object.keys(dependencies).length === 0) {
      return;
    }

    ui.styledHeader(header);
    ui.blankLine();
    ui.table(toPairs(dependencies, { keyName: 'dependency', valueName: 'version' }), {
      dependency: {},
      version: {},
    });
    ui.blankLine();
  }
}
