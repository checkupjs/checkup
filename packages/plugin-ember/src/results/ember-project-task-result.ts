import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class EmberProjectTaskResult extends BaseTaskResult implements TaskResult {
  type!: string;
  name!: string;
  version!: string;

  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
    ui.blankLine();
    ui.styledObject({
      type: this.type,
    });
    ui.blankLine();
  }

  toJson() {
    return {
      [this.meta.taskName]: { type: this.type },
    };
  }
}
