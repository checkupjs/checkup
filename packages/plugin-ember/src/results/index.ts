export { default as EmberProjectTaskResult } from './ember-project-task-result';
export { default as DependenciesTaskResult } from './dependencies-task-result';
export { default as TypesTaskResult } from './types-task-result';
