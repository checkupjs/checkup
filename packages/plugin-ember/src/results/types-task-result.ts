import { TaskItemData, TaskResult, ui } from '@checkup/core';

export default class TypesTaskResult implements TaskResult {
  types!: TaskItemData[];

  toConsole() {
    ui.styledHeader('Types');
    ui.blankLine();
    ui.table(this.types, { type: {}, total: {} });
    ui.blankLine();
  }

  toJson() {
    return { types: this.types };
  }
}
