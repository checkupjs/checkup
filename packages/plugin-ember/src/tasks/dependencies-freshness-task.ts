import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import DependenciesFreshnessTaskResult from '../results/dependencies-freshness-task-result';
import { DepFreshnessInfo } from '../types';

const hash = require('promise.hash.helper');
const shell = require('shelljs');
const OUTDATED_DEP = 'yarn outdated --json';

async function getOutdatedDep(): Promise<DepFreshnessInfo> {
  const { stdout } = await shell.exec(OUTDATED_DEP, { silent: true });

  // stripping out color legend info from the output
  const outdatedDepTable = JSON.parse(stdout.split('}')[1] + '}}');
  // todo: error handle when there is no outdatedDep
  return hash({
    tableHead: outdatedDepTable.data.head,
    tableBody: outdatedDepTable.data.body,
  });
}

export default class DependenciesFreshnessTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'dependencies-freshness',
    friendlyTaskName: 'Dependencies Freshness',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    let result: DependenciesFreshnessTaskResult = new DependenciesFreshnessTaskResult(this.meta);
    result.depFreshnessInfo = await getOutdatedDep();

    return result;
  }
}
