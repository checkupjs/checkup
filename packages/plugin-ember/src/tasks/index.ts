export { default as ProjectInfoTask } from './project-info-task';
export { default as DependenciesTask } from './dependencies-task';
export { default as TypesTask } from './types-task';
