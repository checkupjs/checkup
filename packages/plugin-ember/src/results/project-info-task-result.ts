import { TaskResult, ui } from '@checkup/core';

import { ProjectInfoTask } from '../tasks';

export default class ProjectInfoTaskResult implements TaskResult {
  type!: string;
  name!: string;
  version!: string;

  toConsole() {
    ui.styledHeader(ProjectInfoTask.friendlyTaskName);
    ui.blankLine();
    ui.styledObject({
      name: this.name,
      type: this.type,
      version: this.version,
    });
    ui.blankLine();
  }

  toJson() {
    return {
      [ProjectInfoTask.taskName]: { name: this.name, type: this.type, version: this.version },
    };
  }
}
