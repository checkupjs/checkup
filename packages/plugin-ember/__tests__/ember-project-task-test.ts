import { Project, stdout } from '@checkup/test-helpers';
import { EmberProjectTask } from '../src/tasks';
import { EmberProjectTaskResult } from '../src/results';

describe('project-info-task', () => {
  let fixturifyProject: Project;

  describe('for Ember Applications', () => {
    beforeEach(() => {
      let packageJson = require('./__fixtures__/app-package.json');

      fixturifyProject = new Project('checkup-app', '0.0.0', project => {
        project.addDependency('ember-cli', '^3.15.0');
      });

      fixturifyProject.updatePackageJson(packageJson);

      fixturifyProject.writeSync();
    });

    afterEach(() => {
      fixturifyProject.dispose();
    });

    it('can read project info and output to console', async () => {
      const result = await new EmberProjectTask({ path: fixturifyProject.baseDir }).run();
      const taskResult = <EmberProjectTaskResult>result;

      taskResult.toConsole();

      expect(stdout()).toMatchSnapshot();
    });

    it('can read project info as JSON', async () => {
      const result = await new EmberProjectTask({ path: fixturifyProject.baseDir }).run();
      const taskResult = <EmberProjectTaskResult>result;

      expect(taskResult.toJson()).toMatchSnapshot();
    });
  });

  describe('for Ember Addons', () => {
    beforeEach(() => {
      let packageJson = require('./__fixtures__/addon-package.json');

      fixturifyProject = new Project('checkup-app', '0.0.0', project => {
        project.addDependency('ember-cli', '^3.15.0');
      });

      fixturifyProject.updatePackageJson(packageJson);

      fixturifyProject.writeSync();
    });

    afterEach(() => {
      fixturifyProject.dispose();
    });

    it('can read project info and output to console', async () => {
      const result = await new EmberProjectTask({ path: fixturifyProject.baseDir }).run();
      const taskResult = <EmberProjectTaskResult>result;

      taskResult.toConsole();

      expect(stdout()).toMatchSnapshot();
    });

    it('can read project info as JSON', async () => {
      const result = await new EmberProjectTask({ path: fixturifyProject.baseDir }).run();
      const taskResult = <EmberProjectTaskResult>result;

      expect(taskResult.toJson()).toMatchSnapshot();
    });
  });
});
