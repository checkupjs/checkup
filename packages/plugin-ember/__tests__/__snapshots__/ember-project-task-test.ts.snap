// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project-info-task for Ember Addons can read project info and output to console 1`] = `
"=== Ember Project

type: Ember.js addon

"
`;

exports[`project-info-task for Ember Addons can read project info as JSON 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Ember Project",
    "taskClassification": Object {
      "category": "insights",
      "priority": "medium",
    },
    "taskName": "ember-project",
  },
  "result": Object {
    "type": "Ember.js addon",
  },
}
`;

exports[`project-info-task for Ember Applications can read project info and output to console 1`] = `
"=== Ember Project

type: Ember.js application

"
`;

exports[`project-info-task for Ember Applications can read project info as JSON 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Ember Project",
    "taskClassification": Object {
      "category": "insights",
      "priority": "medium",
    },
    "taskName": "ember-project",
  },
  "result": Object {
    "type": "Ember.js application",
  },
}
`;
